File: python/step1_read_print.py
import sys, traceback
import mal_readline
import mal_types as types
import reader, printer

# read
def READ(str):
    return reader.read_str(str)

# eval
def EVAL(ast, env):
        #print("EVAL %s" % printer._pr_str(ast))
        return ast

# print
def PRINT(exp):
    return printer._pr_str(exp)

# repl
def REP(str):
    return PRINT(EVAL(READ(str), {}))

# repl loop
while True:
    try:
        line = mal_readline.readline("user> ")
        if line == None: break
        if line == "": continue
        print(REP(line))
    except reader.Blank: continue
    except Exception as e:
        print("".join(traceback.format_exception(*sys.exc_info())))

File: ocaml/step1_read_print.ml
let read str = Reader.read_str str
let eval ast any = ast
let print exp = Printer.pr_str exp true
let rep str = print (eval (read str) "")

let rec main =
  try
    while true do
      print_string "user> ";
      let line = read_line () in
        try
          print_endline (rep line);
        with End_of_file -> ()
    done
  with End_of_file -> ()
